[Site Isolation] Fix when about:blank iframes call window.open()

atar13 opened the pull request:
#### 7fcf73d2aff00cac98545c4c11e11c8944bc64af
<pre>
[Site Isolation] Fix when about:blank iframes call window.open()
<a href="https://bugs.webkit.org/show_bug.cgi?id=303956">https://bugs.webkit.org/show_bug.cgi?id=303956</a>
<a href="https://rdar.apple.com/165169147">rdar://165169147</a>

Reviewed by Alex Christensen.

This patch fixes an issue where calling window.open() from the context
of an about:blank iframe. Previously, this snippet would crash with site
isolation enabled.

```
i = document.body.appendChild(document.createElement(&quot;iframe&quot;));
i.contentWindow.open();
```

This patch fixes LayoutTests/http/tests/security/xss-DENIED-synchronous-frame-load-in-javascript-url.html
with site isolation enabled.

This patch changes the site passed into BrowsingContextGroup::ensureProcessForSite
to use the same process as the opener when an about:blank page is opened with
window.open().

Test: http/tests/site-isolation/window-open-from-about-blank-iframe.html

* LayoutTests/http/tests/site-isolation/resources/post-opener-info-message-to-opener.html: Added.
* LayoutTests/http/tests/site-isolation/window-open-from-about-blank-iframe-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/window-open-from-about-blank-iframe.html: Added.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::Internals::Internals):
(WebKit::openerOrigin):
(WebKit::WebPageProxy::WebPageProxy):
(WebKit::WebPageProxy::initializeWebPage):
(WebKit::WebPageProxy::createNewPage):
(WebKit::activityStateChangeTimer): Deleted.
* Source/WebKit/UIProcess/WebPageProxyInternals.h:

Canonical link: <a href="https://commits.webkit.org/304540@main">https://commits.webkit.org/304540@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92a116350682cd996646f40320b2328dc5cef8c8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135278 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7664 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46546 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142783 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/87024 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3f31366b-10db-454f-b351-20ca706596e9) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8300 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7511 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103376 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70798 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/57f2a175-d3ff-435c-a28f-bc875660393d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138224 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5929 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121249 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84239 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5715 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3319 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3370 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114919 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39431 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145479 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7346 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40001 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111755 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7390 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6154 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112122 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5565 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117549 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61276 "The change is no longer eligible for processing.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20927 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7400 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35676 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7154 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70947 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7375 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7257 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

atar13 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -1795,7 +1797,12 @@ void WebPageProxy::initializeWebPage(const Site& site, WebCore::SandboxFlags eff
+
+    // If an empty site openes a new page, this new page should be in the same process
+    // as the opener. To do so, we can pass the opener's origin to BrowsingContextGroup::ensureProcessForSite.
+    Site effectiveSite = site.isEmpty() && m_openerOrigin ? Site { *m_openerOrigin } : site;
```
I was a bit confused on this at first, but in `WebPageProxy::initializeWebPage` `site` represents the site opening the new window (I thought it was the site of the new window). So if we have `window.open` being called from the context of an about:blank iframe (such as `iframe.contentWindow.open("http://localhost:8000")`), then `site` in `initializeWebPage` will be empty (representing about:blank).

atar13 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -905,6 +906,7 @@ WebPageProxy::WebPageProxy(PageClient& pageClient, WebProcessProxy& process, Ref
     , m_menuBarIsVisible(windowFeature([] (auto& features) { return features.menuBarVisible; }, configuration))
     , m_toolbarsAreVisible(windowFeature([] (auto& features) { return features.toolBarVisible; }, configuration)
         || windowFeature([] (auto& features) { return features.locationBarVisible; }, configuration))
+    , m_openerOrigin(configuration->openerInfo().transform([](API::PageConfiguration::OpenerInfo info) { return info.securityOrigin; } ))
```
I'm storing the openerOrigin in WebPageProxy, but doing it during the initializer list before `m_configuration->consumeOpenerInfo()` is called.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -30,6 +30,7 @@
 #include "APIObject.h"
 #include "MessageReceiver.h"
 #include <WebCore/LayerHostingContextIdentifier.h>
+#include <WebCore/SecurityOriginData.h>
```
Please move to WebPageProxyInternals.h instead.

atar13 replied:
Moving to WebPageProxyInternals.h ([diff](https://github.com/WebKit/WebKit/compare/ca8ceb1ac99e55c2656c6a4c037236cac2099cc9..d339ab178e82be1de4abf040a520763de3047536)) caused an issue in Hasher.h
```
/Volumes/Data/worker/macOS-Tahoe-Debug-Build-EWS/build/WebKitBuild/Debug/usr/local/include/wtf/Hasher.h:208:18: error: no member named 'has_value' in 'std::optional<WebCore::SecurityOriginData>'
  208 |     if (optional.has_value())
  ```
  
  Let me see if there's another spot missing `#include <WebCore/SecurityOriginData.h>` or `#include <optional>`

atar13 replied:
Adding `#include <optional>` to WebPageProxyInternals.h doesn't seem to be happy with me using an incomplete type with `std::optional`

```
In template: incomplete type 'WebCore::SecurityOriginData' used in type trait expression
```

atar13 replied:
Everything building after moving openerOrigin to WebPageProxyInternals

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -4092,6 +4092,8 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     bool m_toolbarsAreVisible { true };
     bool m_hasNetworkRequestsInProgress { false };
 
+    std::optional<WebCore::SecurityOriginData> m_openerOrigin;
```
This needs to also be moved to internals.

atar13 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -798,6 +799,7 @@ WebPageProxy::Internals::Internals(WebPageProxy& page)
 #if PLATFORM(GTK) || PLATFORM(WPE)
     , activityStateChangeTimer(RunLoop::mainSingleton(), "WebPageProxy::Internals::activityStateChangeTimer"_s, &page, &WebPageProxy::dispatchActivityStateChange)
 #endif
+    , openerOrigin(openerOrigin)
```
`(WebKit::activityStateChangeTimer): Deleted.`

For some reason, the generated commit message thinks I deleted a method called `activityStateChangeTimer`...

Maybe it has to do with how I ordered the initializer list (since `activityStateChangeTimer` is initialized right before), but I think openerOrigin is in a good spot.

atar13 committed:
[Site Isolation] Fix when about:blank iframes call window.open()
https://bugs.webkit.org/show_bug.cgi?id=303956
rdar://165169147

Reviewed by Alex Christensen.

This patch fixes an issue where calling window.open() from the context
of an about:blank iframe. Previously, this snippet would crash with site
isolation enabled.

```
i = document.body.appendChild(document.createElement("iframe"));
i.contentWindow.open();
```

This patch fixes LayoutTests/http/tests/security/xss-DENIED-synchronous-frame-load-in-javascript-url.html
with site isolation enabled.

This patch changes the site passed into BrowsingContextGroup::ensureProcessForSite
to use the same process as the opener when an about:blank page is opened with
window.open().

Test: http/tests/site-isolation/window-open-from-about-blank-iframe.html

* LayoutTests/http/tests/site-isolation/resources/post-opener-info-message-to-opener.html: Added.
* LayoutTests/http/tests/site-isolation/window-open-from-about-blank-iframe-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/window-open-from-about-blank-iframe.html: Added.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::Internals::Internals):
(WebKit::openerOrigin):
(WebKit::WebPageProxy::WebPageProxy):
(WebKit::WebPageProxy::initializeWebPage):
(WebKit::WebPageProxy::createNewPage):
(WebKit::activityStateChangeTimer): Deleted.
* Source/WebKit/UIProcess/WebPageProxyInternals.h:

Canonical link: https://commits.webkit.org/304540@main
